{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "from targets import create_topological_fasta\n",
    "import argparse\n",
    "import sys\n",
    "from rich.console import Console\n",
    "import os\n",
    "import tempfile\n",
    "from multiprocessing import cpu_count\n",
    "from Bio import SeqIO\n",
    "import re\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "import re\n",
    "import codecs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# consider https://github.com/google/python-fire for command line interface\n",
    "class Args:\n",
    "    genome_file = \"GCA_000005845.2.gb\"\n",
    "    pam = \"NGG\"\n",
    "    barcode_length = 20\n",
    "    orientation = 'reverse'\n",
    "    pam_direction = 'downstream'\n",
    "    omit_intergenic = True\n",
    "    omit_offtargets = True\n",
    "    omit_ambiguous = False\n",
    "    mismatches = 1\n",
    "    keep_top = 10\n",
    "    tile_size = barcode_length\n",
    "    full_overlap = True\n",
    "\n",
    "args = Args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "ESCAPE_SEQUENCE_RE = re.compile(r'''\n",
    "    ( \\\\U........      # 8-digit hex escapes\n",
    "    | \\\\u....          # 4-digit hex escapes\n",
    "    | \\\\x..            # 2-digit hex escapes\n",
    "    | \\\\[0-7]{1,3}     # Octal escapes\n",
    "    | \\\\N\\{[^}]+\\}     # Unicode characters by name\n",
    "    | \\\\[\\\\'\"abfnrtv]  # Single-character escapes\n",
    "    )''', re.UNICODE | re.VERBOSE)\n",
    "\n",
    "def decode_escapes(s):\n",
    "    def decode_match(match):\n",
    "        return codecs.decode(match.group(0), 'unicode-escape')\n",
    "\n",
    "    return ESCAPE_SEQUENCE_RE.sub(decode_match, s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def is_dna(sequence):\n",
    "    return all(base in 'GATC' for base in sequence)\n",
    "\n",
    "def find_sequences_with_barcode_and_pam(topological_fasta_file_name, barcode_length, pam):\n",
    "    matching_sequences = set()\n",
    "    pam_regex = re.compile(pam.replace('N', '[ATGC]'))\n",
    "\n",
    "    with open(topological_fasta_file_name, \"rt\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"fasta\"):\n",
    "            # Consider both the original sequence and its reverse complement\n",
    "            for sequence in [str(record.seq), str(record.seq.reverse_complement())]:\n",
    "                for i in range(len(sequence) - barcode_length - len(pam) + 1):\n",
    "                    # If PAM is downstream\n",
    "                    if args.pam_direction == 'downstream' and pam_regex.match(sequence[i+barcode_length:i+barcode_length+len(pam)]):\n",
    "                        spacer = sequence[i:i+barcode_length]\n",
    "                        if is_dna(spacer):\n",
    "                            matching_sequences.add(spacer)\n",
    "                    # If PAM is upstream\n",
    "                    elif args.pam_direction == 'upstream' and pam_regex.match(sequence[i-len(pam):i]):\n",
    "                        spacer = sequence[i:i+barcode_length]\n",
    "                        if is_dna(spacer):\n",
    "                            matching_sequences.add(spacer)\n",
    "\n",
    "    return matching_sequences\n",
    "\n",
    "# create a sgRNA fasta file such as >sequence\\nsequence\\n\n",
    "def create_sgRNA_fasta(matching_sequences, sgRNA_fasta_file_name):\n",
    "    with open(sgRNA_fasta_file_name, \"wt\") as handle:\n",
    "        for sequence in matching_sequences:\n",
    "            handle.write(f\">{sequence}\\n{sequence}\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(args):\n",
    "    num_threads = cpu_count() // 2\n",
    "    \n",
    "    console=Console(file=sys.stderr)\n",
    "    \n",
    "    console.log(f\"Settings: {args}\")\n",
    "\n",
    "    \n",
    "    with tempfile.TemporaryDirectory() as working_dir:\n",
    "        console.log(\"[bold red]Initializing barcode target builder[/bold red]\")\n",
    "        console.log(f\"Stay tuned... running 'targets.py' to find guides for {args.genome_file} with {args.barcode_length}bp barcodes and {args.pam} PAM sequence\")\n",
    "\n",
    "        topological_fasta_file_name = os.path.join(working_dir, os.path.splitext(os.path.basename(args.genome_file))[0] + \".fasta\")\n",
    "\n",
    "        sgRNA_fasta_file_name = os.path.join(working_dir, \"sgRNA.fasta\")\n",
    "    \n",
    "        create_topological_fasta(args.genome_file, topological_fasta_file_name)\n",
    "    \n",
    "        matching_sequences = find_sequences_with_barcode_and_pam(topological_fasta_file_name, args.barcode_length, args.pam)\n",
    "        \n",
    "        create_sgRNA_fasta(matching_sequences, sgRNA_fasta_file_name)\n",
    "        \n",
    "        result = subprocess.run([\n",
    "            \"python\", \n",
    "            \"targets.py\", \n",
    "            sgRNA_fasta_file_name, \n",
    "            args.genome_file, \n",
    "            args.pam,\n",
    "            \"--pam_direction\", args.pam_direction,\n",
    "            str(args.mismatches)\n",
    "            ], stdout=subprocess.PIPE, text=True, check=True)\n",
    "\n",
    "        targets = pd.read_csv(StringIO(result.stdout), sep='\\t')    \n",
    "        \n",
    "        console.log(f\"Found {len(targets):,} guides\")\n",
    "        console.log(f\"Removing guides based on settings\")\n",
    "        \n",
    "        targets['target'] = targets['target'].str.upper()\n",
    "        \n",
    "        if args.orientation == 'forward':\n",
    "            # omit everything where sp_dir and tar_dir are not the same\n",
    "            targets = targets.loc[targets['sp_dir'] == targets['tar_dir']]\n",
    "        \n",
    "        elif args.orientation == 'reverse':\n",
    "            # omit everything where sp_dir and tar_dir are the same\n",
    "            targets = targets.loc[targets['sp_dir'] != targets['tar_dir']]\n",
    "\n",
    "        if args.omit_offtargets:\n",
    "\n",
    "            console.log(\"[bold red]Removing guides with off-targets[/bold red]\")\n",
    "            len_before = len(targets)\n",
    "            # Extract the number of sites from the 'note' column\n",
    "            targets.loc[:, 'sites'] = targets['note'].str.extract(r'(\\d+) site', expand=False).astype(int)            \n",
    "            # Create a mask that is True for rows where 'sites' is 1\n",
    "            mask = (targets['sites'] == 1)\n",
    "\n",
    "            # Apply the mask to the DataFrame\n",
    "            targets = targets[mask]\n",
    "            console.log(f\"Removed {(len_before - len(targets)):,} guides\")\n",
    "        # remove everything where mismatches > 0\n",
    "        if args.mismatches > 0:\n",
    "            console.log(\"[bold red]Removing guides with mismatches. There shouldn't be any![/bold red]\")\n",
    "            len_before = len(targets)\n",
    "            \n",
    "            targets = targets.loc[targets['mismatches'] == 0]\n",
    "            \n",
    "            console.log(f\"Removed {(len_before - len(targets)):,} guides\")\n",
    "        \n",
    "        if args.omit_ambiguous:\n",
    "            console.log(\"[bold red]Removing ambiguous guides, this will be a lot![/bold red]\")\n",
    "\n",
    "            # Extract the number of sites, genes, and intergenic regions from the 'note' column\n",
    "            targets['sites'] = targets['note'].str.extract(r'(\\d+) site', expand=False).astype(int)\n",
    "            targets['genes'] = targets['note'].str.extract(r'(\\d+) gene', expand=False).astype(int)\n",
    "            targets['intergenic'] = targets['note'].str.extract(r'(\\d+) intergenic', expand=False).astype(int)\n",
    "\n",
    "            # Create a mask that is True for rows where 'sites', 'genes', and 'intergenic' are 1\n",
    "            mask = (targets['sites'] == 1) & (targets['genes'] == 1) & (targets['intergenic'] == 1)\n",
    "\n",
    "            len_before = len(targets)\n",
    "\n",
    "            # Apply the mask to the DataFrame\n",
    "            targets = targets[mask]\n",
    "\n",
    "            console.log(f\"Removed {(len_before - len(targets)):,} guides\")\n",
    "\n",
    "        if args.omit_intergenic:\n",
    "            console.log(\"[bold red]Removing intergenic regions[/bold red]\")\n",
    "\n",
    "            # Create a mask that is True for rows where 'note' does not contain \"intergenic\"\n",
    "            mask = ~targets['note'].str.contains('intergenic')\n",
    "\n",
    "            len_before = len(targets)\n",
    "\n",
    "            # Apply the mask to the DataFrame\n",
    "            targets = targets[mask]\n",
    "\n",
    "            console.log(f\"Removed {(len_before - len(targets)):,} guides\")\n",
    "\n",
    "        \n",
    "        if args.full_overlap:\n",
    "            console.log(\"[bold red]Removing guides that don't fully overlap with the gene[/bold red]\")\n",
    "            # Sort the DataFrame by 'target', 'spacer' and 'locus_tag'\n",
    "            \n",
    "            len_before = len(targets)\n",
    "\n",
    "            # Create a new DataFrame where 'overlap' is equal to args.barcode_length\n",
    "            overlap_df = targets.loc[targets['overlap'] == args.barcode_length]\n",
    "            \n",
    "            # Extract the 'spacer' column from overlap_df and convert it to a set\n",
    "            overlap_spacers = set(overlap_df['spacer'])\n",
    "\n",
    "            # Create a mask that is True for rows where 'spacer' is in overlap_spacers\n",
    "            mask = targets['spacer'].isin(overlap_spacers)\n",
    "\n",
    "            # Apply the mask to the DataFrame\n",
    "            targets = targets[mask]\n",
    "            \n",
    "            console.log(f\"Removed {(len_before - len(targets)):,} guides\")\n",
    "        \n",
    "        # If tile_size was not provided, set it to the value of barcode_length\n",
    "        if args.tile_size is None:\n",
    "            args.tile_size = args.barcode_length\n",
    "            \n",
    "            # Sort the DataFrame by 'locus_tag' and 'offset'\n",
    "            targets = targets.sort_values(['locus_tag', 'offset'])\n",
    "\n",
    "            # Group the DataFrame by 'locus_tag'\n",
    "            grouped = targets.groupby('locus_tag')\n",
    "\n",
    "            # Initialize a set to store the selected spacers\n",
    "            selected_spacers = set()\n",
    "\n",
    "            # Iterate over each group\n",
    "            for name, group in grouped:\n",
    "                last_offset = group['offset'].min() - args.tile_size\n",
    "                for index, row in group.iterrows():\n",
    "                    # If the current offset is at least tile_size away from the last offset, add the spacer to the set\n",
    "                    if row['offset'] >= last_offset + args.tile_size:\n",
    "                        selected_spacers.add(row['spacer'])\n",
    "                        last_offset = row['offset']\n",
    "\n",
    "            # Create a mask that is True for rows where 'spacer' is in selected_spacers\n",
    "            mask = targets['spacer'].isin(selected_spacers)\n",
    "\n",
    "            # Apply the mask to the DataFrame\n",
    "            targets = targets[mask]\n",
    "                        \n",
    "            \n",
    "        if args.keep_top:\n",
    "            console.log(f\"[bold red]Keeping only the top {args.keep_top} guides for each gene[/bold red]\")\n",
    "            \n",
    "            len_before = len(targets)\n",
    "        \n",
    "            # Sort the DataFrame by 'locus_tag', 'offset', and 'overlap'\n",
    "            targets = targets.sort_values(['locus_tag', 'offset', 'overlap'])\n",
    "\n",
    "            # Group the DataFrame by 'locus_tag' and keep the top n rows\n",
    "            top_targets = targets.groupby('locus_tag').head(args.keep_top)\n",
    "\n",
    "            # Get the spacers that are in the top n rows for each gene\n",
    "            top_spacers = top_targets['spacer'].unique()\n",
    "\n",
    "            # Filter the original DataFrame to keep only the rows that are in the top n rows for each gene or that target multiple genes\n",
    "            targets = targets[targets['spacer'].isin(top_spacers)]\n",
    "            \n",
    "            console.log(f\"Removed {(len_before - len(targets)):,} guides\")\n",
    "               \n",
    "    # Convert all numeric columns to integers\n",
    "    targets = targets.apply(lambda col: pd.to_numeric(col, errors='coerce').fillna(0).astype(int) if col.dtypes != object else col)\n",
    "    \n",
    "    targets = targets.sort_values(['chr', 'tar_start', 'tar_end', 'locus_tag', 'offset', 'overlap'])\n",
    "    \n",
    "    # print csv to output\n",
    "    targets.to_csv(sys.stdout, sep='\\t', index=False, na_rep='None')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">[08:55:18] </span>Settings: <span style=\"font-weight: bold\">&lt;</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff; font-weight: bold\">__main__.Args</span><span style=\"color: #000000; text-decoration-color: #000000\"> object at </span><span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0x795377f68590</span><span style=\"font-weight: bold\">&gt;</span>                                        <a href=\"file:///tmp/ipykernel_3927844/198950525.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">198950525.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///tmp/ipykernel_3927844/198950525.py#6\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">6</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[2;36m[08:55:18]\u001b[0m\u001b[2;36m \u001b[0mSettings: \u001b[1m<\u001b[0m\u001b[1;95m__main__.Args\u001b[0m\u001b[39m object at \u001b[0m\u001b[1;36m0x795377f68590\u001b[0m\u001b[1m>\u001b[0m                                        \u001b]8;id=82024;file:///tmp/ipykernel_3927844/198950525.py\u001b\\\u001b[2m198950525.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=354006;file:///tmp/ipykernel_3927844/198950525.py#6\u001b\\\u001b[2m6\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">           </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Initializing barcode target builder</span>                                                      <a href=\"file:///tmp/ipykernel_3927844/198950525.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">198950525.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///tmp/ipykernel_3927844/198950525.py#10\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">10</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[2;36m          \u001b[0m\u001b[2;36m \u001b[0m\u001b[1;31mInitializing barcode target builder\u001b[0m                                                      \u001b]8;id=634149;file:///tmp/ipykernel_3927844/198950525.py\u001b\\\u001b[2m198950525.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=791834;file:///tmp/ipykernel_3927844/198950525.py#10\u001b\\\u001b[2m10\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">           </span>Stay tuned<span style=\"color: #808000; text-decoration-color: #808000\">...</span> running <span style=\"color: #008000; text-decoration-color: #008000\">'targets.py'</span> to find guides for GCA_000005845.<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2.</span>gb with 20bp       <a href=\"file:///tmp/ipykernel_3927844/198950525.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">198950525.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///tmp/ipykernel_3927844/198950525.py#11\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">11</span></a>\n",
       "<span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">           </span>barcodes and NGG PAM sequence                                                            <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[2;36m          \u001b[0m\u001b[2;36m \u001b[0mStay tuned\u001b[33m...\u001b[0m running \u001b[32m'targets.py'\u001b[0m to find guides for GCA_000005845.\u001b[1;36m2.\u001b[0mgb with 20bp       \u001b]8;id=974281;file:///tmp/ipykernel_3927844/198950525.py\u001b\\\u001b[2m198950525.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=678219;file:///tmp/ipykernel_3927844/198950525.py#11\u001b\\\u001b[2m11\u001b[0m\u001b]8;;\u001b\\\n",
       "\u001b[2;36m           \u001b[0mbarcodes and NGG PAM sequence                                                            \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[2;36m[08:55:24]\u001b[0m\u001b[2;36m \u001b[0m\u001b[1;31mInitializing barcode target seeker\u001b[0m                     \u001b]8;id=58277;file:///home/ryandward/Git/barcoder/targets.py\u001b\\\u001b[2mtargets.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=670639;file:///home/ryandward/Git/barcoder/targets.py#452\u001b\\\u001b[2m452\u001b[0m\u001b]8;;\u001b\\\n",
      "\u001b[2;36m          \u001b[0m\u001b[2;36m \u001b[0mAnnotating regions to identify\u001b[33m...\u001b[0m                      \u001b]8;id=699187;file:///home/ryandward/Git/barcoder/targets.py\u001b\\\u001b[2mtargets.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=864182;file:///home/ryandward/Git/barcoder/targets.py#460\u001b\\\u001b[2m460\u001b[0m\u001b]8;;\u001b\\\n",
      "\u001b[2;36m[08:55:29]\u001b[0m\u001b[2;36m \u001b[0mGenerating topological coordinate maps\u001b[33m...\u001b[0m              \u001b]8;id=138137;file:///home/ryandward/Git/barcoder/targets.py\u001b\\\u001b[2mtargets.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=301199;file:///home/ryandward/Git/barcoder/targets.py#474\u001b\\\u001b[2m474\u001b[0m\u001b]8;;\u001b\\\n",
      "\u001b[2;36m[08:55:37]\u001b[0m\u001b[2;36m \u001b[0mAligning annotations to genome\u001b[33m...\u001b[0m                      \u001b]8;id=68044;file:///home/ryandward/Git/barcoder/targets.py\u001b\\\u001b[2mtargets.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=782376;file:///home/ryandward/Git/barcoder/targets.py#479\u001b\\\u001b[2m479\u001b[0m\u001b]8;;\u001b\\\n",
      "\u001b[2;36m[08:55:59]\u001b[0m\u001b[2;36m \u001b[0mFinding matches\u001b[33m...\u001b[0m                                     \u001b]8;id=471723;file:///home/ryandward/Git/barcoder/targets.py\u001b\\\u001b[2mtargets.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=57052;file:///home/ryandward/Git/barcoder/targets.py#486\u001b\\\u001b[2m486\u001b[0m\u001b]8;;\u001b\\\n",
      "\u001b[2;36m[08:56:13]\u001b[0m\u001b[2;36m \u001b[0mAnnotating results\u001b[33m...\u001b[0m                                  \u001b]8;id=517313;file:///home/ryandward/Git/barcoder/targets.py\u001b\\\u001b[2mtargets.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=397315;file:///home/ryandward/Git/barcoder/targets.py#550\u001b\\\u001b[2m550\u001b[0m\u001b]8;;\u001b\\\n",
      "\u001b[2;36m[08:56:22]\u001b[0m\u001b[2;36m \u001b[0mCleaning up\u001b[33m...\u001b[0m                                         \u001b]8;id=61610;file:///home/ryandward/Git/barcoder/targets.py\u001b\\\u001b[2mtargets.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=249233;file:///home/ryandward/Git/barcoder/targets.py#593\u001b\\\u001b[2m593\u001b[0m\u001b]8;;\u001b\\\n"
     ]
    }
   ],
   "source": [
    "main(args)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "barcoder-9Sf8VyPF",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
